### DISCUSSION TITLE
Minimal bit-vector function mode

### DISCUSSION THREAD

**Discussion by chanheec** [verus-lang/verus#76] [Language design]
Similar to `assert_bit_vector`, bit-vector function mode(annotated by `#[verifier(bit_vector)]`) will interpret `requires`, function body expression, and `ensures` as bit-vector clauses, and try to prove it in bit vector theory. After that, it will port the pair of `requires` and `ensures` as integer axiom.

Current `assert_bit_vector` does not recognize variable definitions, and `requires` conditions. Bit-vector function mode will have these two functionalities. Bit-vector function mode might support more functionalities, but at least for now, I plan to only allow a function body that is a pure expression. 

For example, a function that sets two bits in a 32-bit integer would be the follows. 
```
#[verifier(bit_vector)]
#[exec]
fn set_two_bit_exec(bv:u32, low_loc:u32, high:bool, low:bool) -> u32 {
    requires(low_loc < 31);
    ensures(|ret:u32| [ 
        get_bit!(ret, low_loc) == low,
        get_bit!(ret, low_loc+1) == high,
        forall(|loc2:u32| (loc2 < 32 && loc2 != low_loc && loc2 != (low_loc+1)) >>= (get_bit!(ret, loc2) == get_bit!(bv, loc2))),
    ]);
    (bv & (!(3u32 << low_loc))) | (if high {if low {3u32} else {2u32}} else {if low {1u32} else {0u32}} << low_loc)
}

```
Compared to the below current version, which has to call `assert_bit_vector` multiple times, above one would be cleaner. https://github.com/secure-foundations/verus/blob/131a4d3c5876f5c4170e97a0b2d5e6ca77480928/source/docs/bv_expriments/garbage_macro2.rs#L78-L111 

Also, this function mode might be used as `#[proof]` to reduce repeatitiveness when there are some similar calls of `assert_bit_vector`.

**Comment by yizhou7** [comment]
It might be reasonable to have `#[verifier(bit_vector)]` for `proof` mode only with no body. It can then be used in other contexts when needed.

**Comment by chanheec** [comment]
I worked on minimal support for bit-vector function mode(d4ea931b94ed48ab9a9b50104ce03868eb4e289d). Currently below two things are possible.

1. bit-vector proof with no function body (as @yizhou7 suggested above). For example,
https://github.com/secure-foundations/verus/blob/d4ea931b94ed48ab9a9b50104ce03868eb4e289d/source/docs/bv_expriments/garbage.rs#L36-L63


Note that the variable definition in normal mode is repeated in bit-vector mode(line 41). As a result, in line 61, the `set_two_bit_proof`'s precondition(line 41) is checked in `integer mode`.   
  

2. function body with `let` and built-in operators only. For example,
https://github.com/secure-foundations/verus/blob/d4ea931b94ed48ab9a9b50104ce03868eb4e289d/source/docs/bv_expriments/garbage2.rs#L36-L50
